doc/imagewmark.1.md
% IMAGEWMARK(1)	imagewmark-0 | Imagewmark Manual Pages

# NAME
imagewmark - An image program for watermark message bit embedding and extraction


# SYNOPSIS

**imagewmark** {`add`,`get`,`gen-key`} [*OPTIONS*] <*INPUTIMAGE*> [*OUTPUTIMAGE*] [*WATERMARK*]


# DESCRIPTION

**Imagewmark** is a Free Software utility to add an encrypted invisible digital watermark to an image.
Using the same encryption key, the watermark can be reconstructed from cropped, scaled, or compressed
variants of the watermarked image, without knowledge of the original source (blind decoding).
It also includes a command to generate a 128-bit watermarking key that can be used with the `--key` option.

# OPTIONS

Imagewmark supports short and long options which start with two dashes (`--`).

## Global Options

`-h`, `--help`
  : Show a help message about command line usage.

`-v`, `--version`
: Print information about the program version.

# COMMANDS

## Add Watermark
**add** [**OPTIONS**] *INPUT_IMG* *OUTPUT_IMG* *MESSAGE_HEX*

Embed the watermark *MESSAGE_HEX* (up to 128 message bits)
in *INPUT_IMG* and write the result into *OUTPUT_IMG*.

**Command Options:**

`-h`, `--help`
: Show this help message and exit.

`-d`, `--dynamic-zoom` *DYNAMIC_ZOOM*
: Zoom watermark depending on image size (default: -1).

`-z`, `--zoom` *ZOOM*
: Set zoom factor for watermark (default: -1).

`--trace-psnr`
: Compute PSNR (default: False).

`--trace-quality`
: Compute quality metrics (default: False).

`-v`, `--verbose`
: Increase output messages or debugging info for multiple `-vv` (default: 0).

`-q`, `--quiet`
: Reduce output verbosity (default: None).

`-s`, `--strength` *STRENGTH*
: Strength for embedded watermark (default: 2).

`-P` *PLOTS*
: Configure plots with colon separated keywords, 'most' or 'all' (default: None).

`--key` *KEY*
: Load watermarking key from file (default: None).

`--test-key` *TEST_KEY*
: Watermarking key (default: None).

## Get Watermark
**get** [**OPTIONS**] *INPUT_IMG*

Extract a watermark message from the image given as *INPUT_IMG*.

**Command Options:**

`-h`, `--help`
: Show this help message and exit.

`--expect` *EXPECT*
: Expected bit pattern for early exit (default: None).

`--json` [*JSONFILE*]
: Write JSON results into file (default: None).

`--dump` *DUMP*
: Debug flags to dump intermediate stages (default: ).

`--norm-peak-count` *NORM_PEAK_COUNT*
: Number of normalized peaks to use for grid (default: 250).

`--raw-peak-count` *RAW_PEAK_COUNT*
: Number of raw peaks to use for grid (default: 250).

`--jsd-threshold` *JSD_THRESHOLD*
: Jensen-Shannon divergence threshold that is sufficient to stop searching for more watermarks (default: 55%).

`--cornersync` {`on`,`off`,`auto`}
: Set cornersync aided detection mode: 'on' forces it, 'off' disables it, 'auto' is the default.

`--perspective`
: Search for optimal perspective grids (default: False).

`--original` *ORIGINAL*
: Use original image for detection (default: ).

`-v`, `--verbose`
: Increase output messages or debugging info for multiple `-vv` (default: 0).

`-q`, `--quiet`
: Reduce output verbosity (default: None).

`--strength` *STRENGTH*
: Strength for embedded watermark (default: 2).

`-P` *PLOTS*
: Configure plots with colon separated keywords, 'most' or 'all' (default: None).

`--key` *KEY*
: Load watermarking key from file (default: None).

`--test-key` *TEST_KEY*
: Watermarking key (default: None).

## Generate Key Files
**gen-key** [`-h`] [`-v`] [`-q`] *OUTPUT_KEY_FILE*

Generate a 128-bit watermarking key and write the result in hexadecimals into *OUTPUT_KEY_FILE*.

**Command Options:**

`-h`, `--help`
: Show this help message and exit.

`-v`, `--verbose`
: Increase output messages or debugging info for multiple `-vv` (default: 0).

`-q`, `--quiet`
: Reduce output verbosity (default: None).


# EXAMPLES

1. **Generate a 128-bit watermarking key:** \
   Create a private key to encrypt and decrypt the payload.
   ```sh
   imagewmark gen-key mysecret.key
   ```

2. **Embed a watermark message in an image:** \
   Note that watermarks shorter than 128 bits are repeated to fill up to 128 bits:
   ```sh
   imagewmark add [--key mysecret.key] input.png output.png babe
   ```

3. **Extract a watermark message from an image:** \
   Note that the output may contain more than one watermark.

   ```sh
   imagewmark get [--key mysecret.key] output.png
   babebabebabebabebabebabebabebabe JSD=99.94%
   ```


# AUTHORS
This manual page and Imagewmark are written by Tim Janik and Stefan Westerfeld.


# COPYRIGHT
Copyright © 2023-Present The Imagewmark Development Team.
License GPL-3.0+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


# SEE ALSO

[**Imagewmark Website**](https://github.com/tim-janik/imagewmark) \
[**"Local Geometric Distortions Resilient Watermarking Scheme Based on Symmetry"**](https://arxiv.org/abs/2007.10240)
by Zehua Ma, Weiming Zhang, Han Fang, Xiaoyi Dong, Linfeng Geng, Nenghai Yu.


NEWS.md
## Imagewmark NEXT:

### Breaking:
* The default key for Watermark encoding has changed, it was previously unset


## Imagewmark 0.5.0:

### Added:
* Added the --cornersync=<on|off|auto> option to control the corner sync algorithm during extraction
* Added developer documentation, including a description of the architecture
* Added a manual page to the documentation (`imagewmark.1`)
* Added a new tests/README.md to provide an overview and usage instructions for test runs
* Added information about test image datasets to the test documentation
* Added an Alpine CI workflow for automated builds and checks using GitHub Actions
* Added an IRC bot for CI build status notifications
* Added unit tests to verify basic functionality before/after installation
* Added new script (gen-tests-mk) that creates out-of-tree test suites of custom sizes
* Added support to utilize all CPU cores when running test batteries
* Added specific tests for high-resolution images and downscaling attacks

### Breaking:
* We now read all images with OpenCV instead of PIL, this potentially changes the set of supported images
* We assume OpenImageIO/imageio.h and gcrypt.h header files are present
* Dependencies: libgcrypt20-dev libopenimageio-dev libopencv-dev
* Watermark encoding has changed in the 0.5.0 version due to PRNG enum modifications
* We now build PDF versions of the docs only if pdflatex is installed

### Changed:
* Improved test realism by randomizing attack resolutions and shuffled input image selection
* Enhanced test reproducibility with improved and deterministic seeding mechanisms
* Moved all C++ source code into a dedicated cxx/ directory
* Better error handling in build scripts, with improved logging for debugging
* Major overhaul of the build system with unified Makefiles and improved dependency handling
* Fixed various bugs in the build system and C++ code
* Introduced installation check and uninstallation rules (make install, installcheck, uninstall, distcheck)
* Migrated version handling to a shell script with git integration
* Fixed image reading to properly handle subimages and miplevels
* Added check that secret key files are only readable by the user
* We support parallel installations of different major.minor versions

### Removed:
* Removed unused OpenCV configure script
* Removed auto config, compile sources in subdir

### Contributors

Thanks to everyone who made this release happen!

* Tim Janik (@tim-janik)
* Stefan Westerfeld (@swesterfeld)

For full details, see the [commit history](https://github.com/tim-janik/imagewmark/commits/v0.5.0)


## Imagewmark 0.4.0:

* Improved documentation in various places.
* Enhanced test reporting with additional columns, titles, and sorting.
* Extended and separated attack vector generations in the test suite.
* Extended visualizations when following along the detection algorithms.
* Extended performance monitoring by adding CPU time and wall clock measurements to various phases.
* Extended JSON support for reporting detected watermarks.
* Introduced --trace-quality option, to provide scores during embedding.
* Improved test reports, added list generation for failing test cases.
* Fixed test suite crashes, improved test rules and optimized parallel processing.
* Fixed various image processing bugs, removed PIL dependency in favor of CV2 for image IO.
* Use float-to-int rounding conversion for watermark embedder.
* Improve reliability for areas which contain a constant color for some colors.
  - Clip each channel to leave some headroom before embedding the watermark.
* Introduced the imagewmark --ssim option to compute SSIM for images after embedding.
* Improved non-blind decoding via feature based image alignment.
* Improved non-blind decoding via ECC based sub-pixel alignment.
* Improved image extraction performance by relying on corner sync for non-blind decoding.
* Added support for auto-converting CMYK and RGBA images.
* Added support for improved detection via use of the original image (non-blind).
* Add "cornersync" algorithm, to make watermark detection more robust for some attacks:
  - Works if original/wm corners are identical (i.e. jpg, scale, jpg+scale,...).
* Improved peaks2grid for better handling of images with extreme aspect ratios.


## Imagewmark 0.3.0:

* Change strength computation for less visible (less robust) watermarks.
* Improved handling of greyscale images and images in CMYK (tiff format).
* Added experimental perspective virtual grid search using --perspective.
* Improve speed and reduce memory consumption for watermark detection.
  - Pre-scale high resolution images before detection
  - Use 32-bit floats instead of 64-bit complex values for symmetry matrix
  - Compute auto convolution using real FFT
* Robustness improvements:
  - Dynamically adjust zoom for high resolution images
  - Use forward error correction with convolutional codes
  - Try different window sizes for local mean / local variance
* Support fractional zoom for the watermark.
* Improved grid search performance.
* Researched, documented and adjusted statistical threshold for watermark detection.
* Properly documented peaks2grid interface.
* Added JSON output, which provides extracted watermark bits and statistical measures.
* Reworked and added new graphical plot generations, plots can be selected via labels.
* Added a new test suite with maximized parallelism and lots of new attack tests.
* Added new test reports with timing statistics, error percentages, HTML and PDF generation.
* Improved search performance without reducing robustness using JSD match threshold.
* Improved corner case handling and stability.
* Added watermark pattern encryption and decryption via AES key.
* Improved memory usage by releasing auxiliary structures earlier.
* Properly documented new dependencies and improved Python version compatibility.
* Adjusted the Dockerfile to enable fully self contained builds and test runs.


## Imagewmark 0.2.0:


src/imagewmark.py
#!/usr/bin/env python3
# Licensed under the GNU GPL-3.0+: https://www.gnu.org/licenses/gpl-3.0.html

# For benchmarking, start tracking execution time as first thing
from time import time
startup_time = time()

# Load submodules relative to script location
from pathlib import Path
BASE_DIR = Path (__file__).resolve().parent
import sys
sys.path.insert (0, str (BASE_DIR / '../src'))

__version__ = open (str (BASE_DIR / "../.version")).read().strip()

import argparse
import embed, extract
import numpy as np
import scipy.misc, scipy.signal, cv2
import hashlib
from plotting import *
import config, common
import imageio

blurb = """
Script implementing watermark message bit embedding and extraction.
"""

class VerboseAction (argparse.Action):
  def __call__ (self, parser, args, values, option_string=None):
    dir = -1 if option_string.find ("q") >= 0 else +1
    setattr (args, self.dest, dir + getattr (args, self.dest))

# Setup CLI parsr for embedding
def cli_parser():
  p = argparse.ArgumentParser (prog = "imagewmark", description = blurb, formatter_class = argparse.ArgumentDefaultsHelpFormatter)
  a = p.add_argument
  def add_common_args (parser, addget):
    parser.add_argument ('-v', '--verbose', nargs = 0, action = VerboseAction, dest = 'verbose', default = config.verbose,
                         help = "Increase output messages or debugging info for multiple `-vv`")
    parser.add_argument ('-q', '--quiet', nargs = 0, action = VerboseAction, dest = 'verbose',
                         help = "Reduce output verbosity")
    if not addget:
      return
    parser.add_argument ('--strength', type = float, default = 2, help = "Strength for embedded watermark")
    parser.add_argument ('-P', dest = 'plots', type = str, help = "Configure plots with colon separated keywords, 'most' or 'all'")
    parser.add_argument ('--key', type = str, help = "Load watermarking key from file")
    parser.add_argument ('--test-key', type = str, help = "Specify test key (for debugging only, insecure)")

  # COMMANDS
  p.add_argument ('-V', '--version', action = 'version', version = 'ImageWMark %s' % __version__)
  cmdp = p.add_subparsers (title = 'command', dest = 'command', required = True,
                           help = 'Use `<command> -h` for detailed help')
  # ADD
  addp = cmdp.add_parser ('add', help = "Embed a watermark message in an image", formatter_class = argparse.ArgumentDefaultsHelpFormatter)
  addp.add_argument ('input_img', type = str, help = "Image into which the message is to be embedded")
  addp.add_argument ('output_img', type = str, help = "Image with embedded watermark message")
  addp.add_argument ('message_hex', type = str, help = "Message bits to embed (max 128)")
  addp.add_argument ('--dynamic-zoom', type = float, default = -1, help = "Zoom watermark depending on image size")
  addp.add_argument ('--zoom', type = float, default = -1, help = "Set zoom factor for watermark")
  addp.add_argument ('--trace-psnr', action='store_true', help='Compute PSNR')
  addp.add_argument ('--trace-quality', action='store_true', help='Compute quality metrics')
  addp.add_argument ('--py', action='store_true', help='Delegate to Python implementation')
  add_common_args (addp, True)
  # GET
  getp = cmdp.add_parser ('get', help = "Extract a watermark message from an image", formatter_class = argparse.ArgumentDefaultsHelpFormatter)
  getp.add_argument ('input_img', type = str, help = "Image containing embedded watermark message")
  getp.add_argument ('--expect', type = str, help = "Expected bit pattern for early exit")
  getp.add_argument ('--json', dest = 'jsonfile', nargs = '?', default = None, const = '/dev/stdout', help = "Write JSON results into file")
  getp.add_argument ('--dump', type = str, default = '', help = "Debug flags to dump intermediate stages")
  getp.add_argument ('--norm-peak-count', type = int, default = config.norm_peak_count, help = "Number of normalized peaks to use for grid")
  getp.add_argument ('--raw-peak-count', type = int, default = config.raw_peak_count, help = "Number of raw peaks to use for grid")
  getp.add_argument ('--jsd-threshold', type = str, default = config.jsd_threshold, help = "Jensen-Shannon divergence threshold that is sufficient to stop searching for more watermarks")
  getp.add_argument ("--cornersync", choices = ['on', 'off', 'auto'], default = 'auto', help = "Set cornersync aided detection mode: 'on' forces it, 'off' disables it, 'auto' is the default.")
  getp.add_argument ('--perspective', action="store_true", help = "Search for optimal perspective grids")
  getp.add_argument ('--original', type = str, default = '', help = "Use original image for detection")
  add_common_args (getp, True)
  # GEN-KEY
  genp = cmdp.add_parser ('gen-key', help = "Generate 128-bit watermarking key, to be used with the --key option", formatter_class = argparse.ArgumentDefaultsHelpFormatter)
  genp.add_argument ('output_key_file', type = str, help = "File to store watermarking key")
  add_common_args (genp, False)
  return p

# gen-key command
def gen_key (keyfile, args):
  import subprocess, os
  cxx_wmops = Path (__file__).resolve().parent.parent / "cxx" / "wmops"
  proc = subprocess.run ([ cxx_wmops, 'gen-key', keyfile ])
  if proc.returncode:
    raise RuntimeError ("Failed to generate key (%d): %s" % (proc.returncode, ' '.join ([str (cxx_wmops), 'gen-key', keyfile])))

# parse args
args = cli_parser().parse_args()
config.verbose = args.verbose
config.with_debug = args.verbose >= 2
config.enable_plots (args.plots if 'plots' in args else '')
args.startup_time = startup_time
if hasattr (args, 'cornersync'):
  args.cornersync = -1 if args.cornersync.lower() == 'auto' else args.cornersync.lower() == 'on'

# Load Key and prepare Key based PRNGs
if hasattr (args, 'key') or hasattr (args, 'test_key'):
  if not args.test_key and not args.key:
    args.test_key = config.DEFAULT_KEY
  common.load_key (args.key, args.test_key)

# command handling
if args.command == 'add':
  exit_code = 0
  embed.command_add (args.input_img, args.output_img, args.message_hex, args.strength, args)
elif args.command == 'get':
  exit_code = extract.command_get (args.input_img, args.strength, args)
elif args.command == 'gen-key':
  exit_code = gen_key (args.output_key_file, args)
sys.exit (exit_code)
